### PR TITLE
issue-13136 cluster up validates/allows all appropriate docker registry CIDR

### LINKED ISSUE #13136: oc cluster up checks for exact --insecure-registry entry
I have in my /etc/sysconfig/docker an entry with:

````
INSECURE_REGISTRY='--insecure-registry 0.0.0.0/0'
````

which basically makes every registry insecure. "oc cluster up" fails to start as it's checking to an exact configuration entry of "--insecure-registry 172.30.0.0/16"

##### Version
````
oc v1.4.1+3f9807a
kubernetes v1.4.0+776c994
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://172.17.0.1:8443
openshift v1.4.1+3f9807a
kubernetes v1.4.0+776c994
````

### PR DESCRIPTION
**Pull request by jbornemann**
Fixes #13136

@jorgemoralespou @csrwng

### CODE REVIEW

**Comment by csrwng** [comment]
@jbornemann thank you so much for looking into this issue. Just a general comment ... your pull adds support for an unspecified CIDR as with @jorgemoralespou's example. However, I was hoping for a more general solution. For example, if I have `--insecure-registry 172.0.0.0/8`, cluster up should recognize that it does include the CIDR that I'm looking for (172.30.0.0/16). Or if I have 172.28.0.0/15, etc. 

I was thinking of something more along the lines of https://github.com/AndyA/Net--CIDR--Set/blob/master/lib/Net/CIDR/Set.pm (I know, wrong language) ... but basically create a union of all CIDRs specified as --insecure-registry arguments, and check whether the CIDR that we need is a subset of that union.

**Comment by jbornemann** [comment]
Hey @csrwng! Thanks for the quick feedback!

I've added some changes. Wondering if we can utilize net.IPNet.Contains() to do some of the lifting for us. Also started adding some tests - will pick it back up soon.

**Comment by csrwng** [comment]
@jbornemann thanks for updating the code. net.IPNet.Contains will only check whether a single IP address is contained within the subnet and net.IPNet.IP will only return the single IP address 172.30.0.0 or whatever you used to specify your CIDR which I don't think quite does it. The CIDR may not include the entire range of addresses.

**Comment by jbornemann** [comment]
Hey @csrwng, you are absolutely right! A prime example being 172.30.0.0/24. 

What I have done is, for a given set of registry entries - I check the number of bits in each netmask, and throw out any that are greater than the number of bits in the netmask of the CIDR we are comparing against. Reason being, any registry CIDR that at least contains the ip range of the CIDR we are interested in should contain the same or greater number of potential IPs. I then see if the candidate network number fits in the remaining "union" of registry IP ranges. 

Finished up some tests. Let me know if there is anything I am missing, fundamental or otherwise! I appreciate the help!

**Comment by csrwng** [comment]
@jbornemann sorry for the long delay in looking at this. I will review today.

**Comment by jbornemann** [comment]
No problem! Happy for your help

**Comment by csrwng** [comment]
@jbornemann thanks for the updated code. I don't think that it's entirely correct to throw out any entries that are not the size of the subnet that you are looking for. For example, if I have --insecure-registry arguments with these CIDRs: `172.30.0.0/18`, `172.30.64.0/18`, and `172.30.128.0/17`, when you get their union, they do match `172.30.0.0/16`.

For something like this, it may be helpful to think of the IPv4 address space as the set of unsigned 32-bit integers. In the end, that's all an IPv4 address is. If you then think of a CIDR as a range of integers with a lower bound and an upper bound, then doing operations like unions and testing whether one contains another is pretty simple.

I wrote some code that demonstrates this. Of course it doesn't include any tests :-) Feel free to make it your own. 

Also I don't think you need to special case the unspecifiedCIDR ... that's a CIDR like any other.

https://gist.github.com/csrwng/5191d707cf3cfa4dcac93ce90674f84d

**Comment by jbornemann** [comment]
AH, I see what you are saying. I didn't know Docker evaluated the registry entries as a union that way. I always assumed they were evaluated in a mutually exclusive fashion. 

I'll get back to this at some point soon.

**Comment by csrwng** [comment]
> I didn't know Docker evaluated the registry entries as a union that way.

Just to clarify ... Docker will only check whether a single IP (when trying to push to it) is contained by any of the CIDRs specified as a --insecure-registry argument. However, it is an OpenShift requirement that any IP within the 172.30.0.0/16 cidr be allowed to pass Docker's insecure-registry check. If the union of --insecure-registry arguments meets this requirement, then we should allow it.

**Comment by jbornemann** [comment]
Hey @csrwng, thanks for the guidance! I made changes based on your feedback. Take a look when you have a chance

File: pkg/bootstrap/docker/dockerhelper/ipv4range.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package dockerhelper
+
+import (
+    "sort"
+    "net"
+)
+
+type ipv4 uint32
+
+type IPV4Range struct {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/ipv4range.go]: Add godoc

File: pkg/bootstrap/docker/dockerhelper/ipv4range.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    from ipv4
+    to ipv4
+}
+
+func (r IPV4Range) Contains(aRange IPV4Range) bool {
+    return aRange.from >= r.from && aRange.to <= r.to
+}
+
+type IPV4RangeList []IPV4Range
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/ipv4range.go]: Add godoc

File: pkg/bootstrap/docker/dockerhelper/ipv4range.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+type ipv4 uint32
+
+type IPV4Range struct {
+    from ipv4
+    to ipv4
+}
+
+func (r IPV4Range) Contains(aRange IPV4Range) bool {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/ipv4range.go]: Add godoc

File: pkg/bootstrap/docker/dockerhelper/registryconfig.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package dockerhelper
+
+import (
+	"net"
+
+	"github.com/docker/engine-api/types"
+	"github.com/golang/glog"
+)
+
+type RegistryConfig struct {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/registryconfig.go]: Add godoc

File: pkg/bootstrap/docker/dockerhelper/registryconfig.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	ipv4Ranges IPV4RangeList
+}
+
+func NewRegistryConfig(dockerInfo *types.Info) (*RegistryConfig) {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/registryconfig.go]: Add godoc

File: pkg/bootstrap/docker/dockerhelper/ipv4range.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    return result
+}
+
+func toIPv4(ip net.IP) ipv4 {
+    ip = ip.To4()
+    if len(ip) < 4 {
+        return 0
+    }
+    return  (ipv4)(
+                uint32(ip[0]) << 24 |
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/ipv4range.go]: 👍

File: pkg/bootstrap/docker/dockerhelper/ipv4range.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+func fromCIDR(ipnet *net.IPNet) IPV4Range {
+    // The lower bound of the range is
+    // always going to be the IP associated with
+    // the IPNet
+    lower := toIPv4(ipnet.IP)
+
+    // The upper bound can be calculated by adding
+    // the inverse of the mask to the lower bound.
+    ones, _ := ipnet.Mask.Size()
+    toAdd := 0XFFFFFFFF >> uint32(ones)
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/ipv4range.go]: 👍

File: pkg/bootstrap/docker/dockerhelper/registryconfig.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	ipv4Ranges := make(IPV4RangeList, 0)
+	for _, cidr := range dockerInfo.RegistryConfig.InsecureRegistryCIDRs {
+		ipv4Ranges = append(ipv4Ranges, fromCIDR((*net.IPNet)(cidr)))
+	}
+	return &RegistryConfig{ipv4Ranges}
+}
+
+func (config *RegistryConfig) HasCustomInsecureRegistryCIDRs() bool {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/registryconfig.go]: godoc

File: pkg/bootstrap/docker/dockerhelper/registryconfig.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	count := len(config.ipv4Ranges)
+	glog.V(5).Infof("Contains %d --insecure-registry entries", count)
+	//Docker always includes 127.0.0.0/8
+	return count > 1
+}
+
+func (config *RegistryConfig) ContainsInsecureRegistryCIDR(cidr string) (bool, error) {
```
Comment (csrwng) [inline pkg/bootstrap/docker/dockerhelper/registryconfig.go]: godoc

File: pkg/oc/bootstrap/docker/dockerhelper/helper.go [hunk 10]
```diff
@@ -81,18 +64,22 @@ func (h *Helper) CgroupDriver() (string, error) {
 	return info.CgroupDriver, nil
 }
 
-// HasInsecureRegistryArg checks whether the docker daemon is configured with
-// the appropriate insecure registry argument
-func (h *Helper) HasInsecureRegistryArg() (bool, error) {
+//Checks to see if the Docker daemon has an appropriate insecure registry argument set so that our services can access the registry
```
Comment (csrwng) [inline pkg/oc/bootstrap/docker/dockerhelper/helper.go]: Start godoc with function name

File: pkg/oc/bootstrap/docker/up.go [hunk 11]
```diff
@@ -647,13 +647,16 @@ func (c *CommonStartConfig) CheckOpenShiftImage(out io.Writer) error {
 	return c.DockerHelper().CheckAndPull(c.openshiftImage(), out)
 }
 
-// CheckDockerInsecureRegistry checks whether the Docker daemon is using the right --insecure-registry argument
+//Checks to see if the Docker daemon has an appropriate insecure registry argument set so that our services can access the registry
```
Comment (csrwng) [inline pkg/oc/bootstrap/docker/up.go]: Start godoc with function name

**Comment by jbornemann** [comment]
@csrwng thank you so much for your review. I've updated documentation, run fmt, and rebased.

**Comment by csrwng** [comment]
/lgtm
thx!

**Pull request merged**